### ISSUE TITLE
Feature Request: Using A Conditional Type With Never In A Generic Type Should Act As A Type Guard

### ISSUE THREAD

**Issue by Daniihh** [microsoft/TypeScript#35507]
### Code
```typescript
type MyCoolType<T> =
    T extends number ? {secretStuff: any} : {new(): any};

declare function getMyCoolType<T>(input: T): MyCoolType<T>;

function operateOnMycoolType<T>(input: T) {
    let output = getMyCoolType(input);
    if ("secretStuff" in output) return;

    let Constructor = output as MyCoolType<T extends number ? never : T>;
    new Constructor(); //Should not complain that this type may not have
    //a construct signature, because we already specified that T being a
    //number will never happen.
}
```

### Expected Behavior
By specifying that `T` is `never` a `number`, TypeScript shouldn't have a problem constructing the type `MyCoolType`, because `MyCoolType` doesn't have a construct signature *only* when `T` is a `number`.

### Actual Behavior
TypeScript still thinks `MyCoolType<T extends number ? never : T>` might not always have a constructor signature.

### Other Information
If this result is expected and this isn't a bug, then I'd like this to be a feature request.
TypeScript version: 3.5.3

**Comment by RyanCavanaugh** [comment]
It would be really great to see a more real-world-like example that shows why you'd end up in this state.

**Comment by Benjamin-Dobell** [comment]
I _think_ there may be two issues here:

 1. `never` being distributive across conditional types. (Known, but inaccurate, behavior).
 2. A misreporting of the source type responsible for the "has no construct signatures" error.

I can't prove the second issue, however, I can prove that the first issue does at least lead to a "has no construct signatures" on `never` error, as follows.

# Proof

**Utilities**
```typescript
declare function returnNever(): never;
```
_**Note**: This is not possible to implement, but the type system doesn't know that._

First we show that `operateOnMycoolType<T>(input: T)` accepts `number`as `T`:

```typescript
operateOnMycoolType<numeber>(returnNever())
```
_**Note**: `never` is also accepted and has the same outcome._ 

Show that `MyCoolType<T>` accepts `T = never`, and that the resulting type is `never` i.e. Issue 1 above (`never` is distributive across conditional types, resulting in `never`):

```typescript
const myCoolType: MyCoolType<never> = returnNever()
```

i.e. The inline cast can result in type `never`:

```typescript
type InlineCastType<T> = MyCoolType<T extends number ? never : T> // From line 10
const Constructor: InlineCastType<number> = returnNever()
```

_**Note**: That Constructor is type `never` *not* `{secretStuff: any}`_

Show that never (of course) does not have any construct signatures:

```typescript
new Constructor() // Type 'never' has no construct signatures
```

[Playground](https://www.typescriptlang.org/play/index.html#code/PTAEDECcHsFtQBYBckAcDOAuEBzAlkggK4BGAdAMZzCx4UzrQBmSwAKgJ6oCmAyvXlSs86dEW7pgAZgCsMgAwB2AFAhQAXk2blypF26gAshwDC0aABtOPADxsAfBuWgXoNqG4APJNwB2AE3RQXyJYEm5IUAB+UABvdG56biReJCImJkxQAENfDgBfUCzY324AdwAKAEos3IKAbh1-RItsyAMmIl8KJDxoX1AcZOMzS2tuO3sKvF9UIiQsthqjU3MrfUnG5U7u3v7QaB5IbJ8AeV9jKjGNh2nZ+cWquOdXC2SD+bmkDUHh1eueHcvlVGq5QHgmKAKgAiBJJFJpDLQ8EDaCfeZPdppSC+LZgt7fMy+dBISBEHrQSLqD5oeY5IIjNbjOwebx+QLBULhSIxUoANwiRTc9lBrlKZVARJJZIpkGq9VAIF4CDRFn8wWg3yosFQrRmoEIJwNCBEBv0oFg2Q4Gu+CGyApeLhA2VAVGJpPJ33QeBwvhORHaABpQOEKNkiAlQGUDNkLO1sv5regeBQIXhuOrDd93OEZjgco7FcAQmFBWU8BYLMFuALInbUDxfGRlPlVGAtNplM0KK12qAdj0+gMsQHfAA5GsRapZfkRLZqZXQCVZg5HE7cc6XJk3KYzL6PHIUCjcIRBdzUkvc5SHCLrzccK7rWyXiJTEc4ie16pVHQLlXLu1vkZAEJgcQ9j1PNwflnSBg1yTNAONAx2jECxel8fM9B4cEghg5s3RJC1-ifbgsmAkibBgxxqXfcdJzlH823BMhuGbLCDAASV8CwZm4ExshJZkwOpcihNZHwAlwrlBV5eihTAtQoDgUAeNKUAAEYpGUAjCX6aVPUpLIuNUviBKQZkX0gajQFoz8p0YtQx01Ui3EQqUPVlHCzWwgADGCfNAAAqXxNUCuI4SxVJ0kyHI8lbJjFwAo0YKhZhXTRSAEiefxoAkG1EHtGN0vdGUkEpUBvV9f12nc0rKWUcVJT0jyyoYos3HNAByGDOoK3DoFdWMq0qv1sW4ZQgA)

# Root cause

I haven't _at all_ proved issue 2 here, however, even if that doesn't hold up to scrutiny. The code in question isn't going to type check successfully due to the above. 

However, I _believe_, the underlying issue has been reported a few times under many many guises. To my knowledge the first report was in #23182.

i.e. that `never` is redistributed across conditional types.

Unfortunately, whilst presently being accepted as by design, this is mathematically unsound behaviour.

i.e.

```typescript
type NumberOrNever<T> = T extends number ? number : number
let supposedToBeANumber: NumberOrNever<never>
supposedToBeANumber = 1 // Type `1` is not assignable to type `never`
supposedToBeANumber = returnNever() // Assigns i.e. supposedToBeANumber is indeed `never`
```

The justification for the above is that `never` represents the so called _empty union_. Thus when you distribute an _empty union_, you get an _empty union_.

The issue is that, in logic / mathematics (i.e. set theory), there's no such concept as an _empty union_.

It's instead a concept introduced by programmers to simplify their code. Instead, mathematically a union is a binary operation, _it requires two operands_.

As such, `never` could be at best defined as the union `never` **U** `never`.

```typescript
type never = never | never
```

Of course, that can't possibly be the real definition of `never`, as it's recursive!

The real definition for `never` is simply the "empty set".

If you _really_ want to define that in terms of set operations (without resorting to recursive hocus pocus) it can be defined as the _intersection of any two disjoint sets_.

e.g.

```typescript
type never = string & number
```

Sadly, it looks like this behaviour is here to stay.

**Comment by Daniihh** [comment]
Thanks @Benjamin-Dobell for the great explanation about what is happening here. And yes, @RyanCavanaugh, it is a request for negative types in a sense.

I'm going to close this issue now that @Benjamin-Dobell brought to my attention issue #23182, which was closed with "Working as Intended" and there is no need for this issue to be open. Sorry for dragging this issue on for over 3 months.
